engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -172,78 +232,211 @@ private void validate(SnapshotInfo snapshotInfo) {
+                }
+
+                Map<String, String> srcDetails = getSnapshotDetails(snapshotInfo);
+
+                copyCommand.setOptions(srcDetails);
+
+                copyCmdAnswer = (CopyCmdAnswer)_agentMgr.send(hostVO.getId(), copyCommand);
             }
             catch (Exception ex) {
-                s_logger.debug(ex.getMessage(), ex);
+                throw new CloudRuntimeException(ex.getMessage());
             }
+            finally {
+                try {
+                    _volumeService.revokeAccess(snapshotInfo, hostVO, srcDataStore);
+                }
+                catch (Exception ex) {
+                    s_logger.debug(ex.getMessage(), ex);
+                }
 
-            if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
-                if (copyCmdAnswer != null && copyCmdAnswer.getDetails() != null && !copyCmdAnswer.getDetails().isEmpty()) {
-                    errMsg = copyCmdAnswer.getDetails();
+                if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
+                    if (copyCmdAnswer != null && !StringUtils.isEmpty(copyCmdAnswer.getDetails())) {
+                        errMsg = copyCmdAnswer.getDetails();
+                    }
+                    else {
+                        errMsg = "Unable to perform host-side operation";
+                    }
                 }
-                else {
-                    errMsg = "Unable to perform host-side operation";
+
+                try {
+                    if (errMsg == null) {
+                        snapshotInfo.processEvent(Event.OperationSuccessed);
+                    }
+                    else {
+                        snapshotInfo.processEvent(Event.OperationFailed);
+                    }
+                }
+                catch (Exception ex) {
+                    s_logger.debug(ex.getMessage(), ex);
                 }
             }
 
-            try {
-                if (errMsg == null) {
-                    snapshotInfo.processEvent(Event.OperationSuccessed);
+            CopyCommandResult result = new CopyCommandResult(null, copyCmdAnswer);
+
+            result.setResult(errMsg);
+
+            callback.complete(result);
+        }
+        finally {
+            if (usingBackendSnapshot) {
+                deleteVolumeFromSnapshot(snapshotInfo);
+            }
+        }
+    }
+
+    /**
+     * Clones a template present on the storage to a new volume and resignatures it.
+     *
+     * @param srcData   source template
+     * @param destData  destination ROOT volume
+     * @param callback  for async
+     */
+    private void handleCreateVolumeFromTemplateBothOnStorageSystem(DataObject srcData, DataObject destData, AsyncCompletionCallback<CopyCommandResult> callback) {
+        CopyCmdAnswer copyCmdAnswer = null;
+        String errMsg = null;
+
+        TemplateInfo templateInfo = (TemplateInfo)srcData;
+        VolumeInfo volumeInfo = (VolumeInfo)destData;
```

Comment:
@DaanHoogland the ``checkArgument`` calls protects against an NPE, but do not check the types to avoid a ``ClassCastException``.  Adding ``Preconditions.checkArgument(srcData instanceof TemplateInfo, "<useful error message>")`` and ``Preconditions.checkArgument(destData instanceof VolumeInfo, "<useful error message>")``.